### ISSUE TITLE
wrong associative array merge on services

### ISSUE THREAD

**Issue by blaggacao** [GoogleContainerTools/kpt#1143]
```console
$ kpt version
0.34.0

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"archive", BuildDate:"1980-01-01T00:00:00Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6+k3s1", GitCommit:"6f56fa1d68a5a48b8b6fdefa8eb7ead2015a4b3a", GitTreeState:"clean", BuildDate:"2020-07-16T20:46:15Z", GoVersion:"go1.13.11", Compiler:"gc", Platform:"linux/amd64"}
```

`kpt pkg update --strategy resource-merge` ***wrongly*** yields me

```diff
kind: Service
metadata:
  name: aedir
spec:
  ports:
  - name: ldap-tcp
    protocol: TCP
    port: 389
    targetPort: ldap
    # appProtocol: ldap  # fg: ServiceAppProtocol
-  - name: ldap-udp
-    protocol: UDP
+  - name: ldap-tcp
+    protocol: TCP
    port: 389
    targetPort: ldap
    # appProtocol: ldap  # fg: ServiceAppProtocol
```

from this upstream

```yaml
apiVersion: v1
kind: Service
metadata:
  name: aedir
spec:
  ports:
  - name: ldap-tcp
    protocol: TCP
    port: 389
    targetPort: ldap
    # appProtocol: ldap  # fg: ServiceAppProtocol
  - name: ldap-udp
    protocol: UDP
    port: 389
    targetPort: ldap
    # appProtocol: ldap  # fg: ServiceAppProtocol
```

I suspect the `name` attribute is not properly prioritized for associative array merge within services.

**Comment by mortent** [comment]
This is related to https://github.com/kubernetes-sigs/kustomize/issues/3111 which is currently being worked on. Once that is fixed in the underlying `kyaml` library it should work in kpt too.

**Comment by natasha41575** [comment]
@monopole

**Comment by apelisse** [comment]
Yeah, that's because of the protocol not being part of the key. Exactly the problem we're trying to solve, thanks!

**Comment by blaggacao** [comment]
Let me say this: using and looping over `kpt` and `kustomize` is a very rewarding experience. Thank you all!

**Comment by natasha41575** [comment]
This should be fixed by #3159 (see issue #3111), which should change the merge key for service ports to be {port, protocol} according to the openapi schema. But it will still not allow port aliasing if the port has the same port number and protocol.

**Comment by natasha41575** [comment]
kubernetes-sigs/kustomize/pull/3159 has been merged so this issue is fixed in the underlying kyaml library
